qiime2/sdk/result.py

```diff
@@ -282,6 +286,205 @@ def result(self):
         """
         return self
 
+    def _validate_annotation_support(self):
+        """Checks for the existance of `annotations_dir` on a Result's
+        format class to guard against annotation actions being called on
+        Results with versions < 7.0.
+
+        Raises
+        ------
+        ValueError\n
+            If the Result's format class has no `annotations_dir` and is
+            thus a format version < 7.0.
+
+        """
+        if self._archiver.annotations_dir is None:
+            raise ValueError(
+                'The Artifact or Visualization being used is associated with '
+                'a QIIME 2 archive format of < 7.0. '
+                'Annotation actions are only supported for QIIME 2 archive '
+                'formats of 7.0 and above.'
+            )
+
+    def _read_annotations(self):
+        """
+        Append any existing Annotations to `self._annotations`.
+        Helper method for `add_annotation`, after a given Annotation
+        has been written to disk.
+        """
+        self._annotations = []
+        annotations_dir = self._archiver.annotations_dir
+        # Not sure if there's a better way to do this, but this accounts for
+        # the fact that the annotations_dir will be None for all
+        # previous archive versions < 7.0
+        if annotations_dir and os.path.exists(annotations_dir):
+            for annotation in os.listdir(annotations_dir):
+                self._annotations.append(
+                    Annotation.load(os.path.join(annotations_dir, annotation))
+                )
+
+    def add_annotation(self, annotation):
+        """Add an Annotation onto a Result object.
+        All Result-associated parameters are passed into the sub-class's
+        `write` method, while the Annotation instance handles everything else.
+
+        Parameters
+        ----------
+        annotation\n
+            An instantiated Annotation subclass (Note, etc).
+
+        Raises
+        ------
+        ValueError\n
+            If the Annotation name matches an existing Annotation name
+            attached to the Result in question.
+
+        Notes
+        -----
+            In Archive Format 7.0, `referenced_result_uuid` is set to
+            the same value as `root_result_uuid`, but this will change
+            in future versions to allow for Annotations that may reference
+            a different Result than the one they are attached to.
+
+        See Also
+        --------
+        Annotation.write
```

Comment:
unless you want users calling `Annotation.write` directly, I would leave this out and say nothing about it

Reply:
updated the whole docstring as a comment, so just leaving this as-is for us to reference in the future